File: 61-rotate-list/rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    if (!head) {
        return head;
    }

    let length = 1;
    let tail = head;

    while (tail.next) {
        tail = tail.next;
        length++;
    }

    k = k % length;
    if (k == 0) {
        return head;
    }

    let cur = head;
    for (let i = 0; i < length - k - 1; i++) {
        cur = cur.next;
    }
    let newHead = cur.next;
    cur.next = null;
    tail.next = head;
    return newHead;
};

File: 61-rotate-list/rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode temp = head;
        int length = 1;

        while (temp.next != null) {
            temp = temp.next;
            length++;
        }

        int actualK = k % length;

        if (actualK == 0) {
            return head;
        }

        ListNode breakNode = head;
        ListNode newHead;
        for (int i = 0; i < length - actualK - 1; i++) {
            breakNode = breakNode.next;
        }

        newHead = breakNode.next;
        breakNode.next = null;
        temp.next = head;

        return newHead;
    }
}